### DISCUSSION TITLE
CSLA 6 prerelease 4 - exception thrown in DataPortal App-Server - asking here before raising a ticket

### DISCUSSION THREAD

**Discussion by vijaygill** [MarimerLLC/csla#2729] [Questions] [answered]
Hi

I upgraded to latest (pre-release) nuget packages and I get following exception at startup of the dataportal app server. The code is from the book CSLA-Blazor-200519 (Chapter 7, section Server-side Blazor with App Server) and has been working fine till prerelease 3. Since there have been quite a few changes in prerelease 4, clearly I am missing something. In the startup.cs of app-server, CSLA is added by "services.AddCsla();".

**Exception thrown:**

```
System.AggregateException
  HResult=0x80131500
  Message=Some services are not able to be constructed (Error while validating the service descriptor 'ServiceType: Csla.ApplicationContext Lifetime: Scoped ImplementationType: Csla.ApplicationContext': Unable to resolve service for type 'Microsoft.AspNetCore.Components.Authorization.AuthenticationStateProvider' while attempting to activate 'Csla.Blazor.ApplicationContextManager'.) (Error while validating the service descriptor 'ServiceType: Csla.Core.IContextManager Lifetime: Scoped ImplementationType: Csla.Blazor.ApplicationContextManager': Unable to resolve service for type 'Microsoft.AspNetCore.Components.Authorization.AuthenticationStateProvider' while attempting to activate 'Csla.Blazor.ApplicationContextManager'.) (Error while validating the service descriptor 'ServiceType: Csla.Channels.Local.LocalProxy Lifetime: Transient ImplementationType: Csla.Channels.Local.LocalProxy': Unable to resolve service for type 'Microsoft.AspNetCore.Components.Authorization.AuthenticationStateProvider' while attempting to activate 'Csla.Blazor.ApplicationContextManager'.) (Error while validating the service descriptor 'ServiceType: Csla.Server.IAuthorizeDataPortal Lifetime: Scoped ImplementationType: Csla.Server.ActiveAuthorizer': Unable to resolve service for type 'Microsoft.AspNetCore.Components.Authorization.AuthenticationStateProvider' while attempting to activate 'Csla.Blazor.ApplicationContextManager'.) (Error while validating the service descriptor 'ServiceType: Csla.Server.IInterceptDataPortal Lifetime: Scoped ImplementationType: Csla.Server.Interceptors.ServerSide.RevalidatingInterceptor': Unable to resolve service for type 'Microsoft.AspNetCore.Components.Authorization.AuthenticationStateProvider' while attempting to activate 'Csla.Blazor.ApplicationContextManager'.) (Error while validating the service descriptor 'ServiceType: Csla.Server.IObjectFactoryLoader Lifetime: Scoped ImplementationType: Csla.Server.ObjectFactoryLoader': Unable to resolve service for type 'Microsoft.AspNetCore.Components.Authorization.AuthenticationStateProvider' while attempting to activate 'Csla.Blazor.ApplicationContextManager'.) (Error while validating the service descriptor 'ServiceType: Csla.Server.IDataPortalActivator Lifetime: Scoped ImplementationType: Csla.Server.DefaultDataPortalActivator': Unable to resolve service for type 'Microsoft.AspNetCore.Components.Authorization.AuthenticationStateProvider' while attempting to activate 'Csla.Blazor.ApplicationContextManager'.) (Error while validating the service descriptor 'ServiceType: Csla.Server.IDataPortalServer Lifetime: Transient ImplementationType: Csla.Server.DataPortal': Unable to resolve service for type 'Microsoft.AspNetCore.Components.Authorization.AuthenticationStateProvider' while attempting to activate 'Csla.Blazor.ApplicationContextManager'.) (Error while validating the service descriptor 'ServiceType: Csla.Server.InterceptorManager Lifetime: Scoped ImplementationType: Csla.Server.InterceptorManager': Unable to resolve service for type 'Microsoft.AspNetCore.Components.Authorization.AuthenticationStateProvider' while attempting to activate 'Csla.Blazor.ApplicationContextManager'.) (Error while validating the service descriptor 'ServiceType: Csla.Server.DataPortalSelector Lifetime: Transient ImplementationType: Csla.Server.DataPortalSelector': Unable to resolve service for type 'Microsoft.AspNetCore.Components.Authorization.AuthenticationStateProvider' while attempting to activate 'Csla.Blazor.ApplicationContextManager'.) (Error while validating the service descriptor 'ServiceType: Csla.Server.SimpleDataPortal Lifetime: Transient ImplementationType: Csla.Server.SimpleDataPortal': Unable to resolve service for type 'Microsoft.AspNetCore.Components.Authorization.AuthenticationStateProvider' while attempting to activate 'Csla.Blazor.ApplicationContextManager'.) (Error while validating the service descriptor 'ServiceType: Csla.Server.FactoryDataPortal Lifetime: Transient ImplementationType: Csla.Server.FactoryDataPortal': Unable to resolve service for type 'Microsoft.AspNetCore.Components.Authorization.AuthenticationStateProvider' while attempting to activate 'Csla.Blazor.ApplicationContextManager'.) (Error while validating the service descriptor 'ServiceType: Csla.Server.DataPortalBroker Lifetime: Transient ImplementationType: Csla.Server.DataPortalBroker': Unable to resolve service for type 'Microsoft.AspNetCore.Components.Authorization.AuthenticationStateProvider' while attempting to activate 'Csla.Blazor.ApplicationContextManager'.) (Error while validating the service descriptor 'ServiceType: Csla.DataPortalClient.IDataPortalProxy Lifetime: Transient ImplementationType: Csla.Channels.Local.LocalProxy': Unable to resolve service for type 'Microsoft.AspNetCore.Components.Authorization.AuthenticationStateProvider' while attempting to activate 'Csla.Blazor.ApplicationContextManager'.)
  Source=Microsoft.Extensions.DependencyInjection
  StackTrace:
   at Microsoft.Extensions.DependencyInjection.ServiceProvider..ctor(ICollection`1 serviceDescriptors, ServiceProviderOptions options)
   at Microsoft.Extensions.DependencyInjection.ServiceCollectionContainerBuilderExtensions.BuildServiceProvider(IServiceCollection services, ServiceProviderOptions options)
   at Microsoft.Extensions.DependencyInjection.DefaultServiceProviderFactory.CreateServiceProvider(IServiceCollection containerBuilder)
   at Microsoft.Extensions.Hosting.Internal.ServiceFactoryAdapter`1.CreateServiceProvider(Object containerBuilder)
   at Microsoft.Extensions.Hosting.HostBuilder.CreateServiceProvider()
   at Microsoft.Extensions.Hosting.HostBuilder.Build()
   at BlazorBusinessApp.DataPortalServer.Program.Main(String[] args) in C:\Temp\Dev\Blazor\BlazorBusinessApp\BlazorBusinessApp.DataPortalServer\Program.cs:line 16
```

  This exception was originally thrown at this call stack:

```
    [External Code]

Inner Exception 1:
InvalidOperationException: Error while validating the service descriptor 'ServiceType: Csla.ApplicationContext Lifetime: Scoped ImplementationType: Csla.ApplicationContext': Unable to resolve service for type 'Microsoft.AspNetCore.Components.Authorization.AuthenticationStateProvider' while attempting to activate 'Csla.Blazor.ApplicationContextManager'.

Inner Exception 2:
InvalidOperationException: Unable to resolve service for type 'Microsoft.AspNetCore.Components.Authorization.AuthenticationStateProvider' while attempting to activate 'Csla.Blazor.ApplicationContextManager'.
```

**Comment by rockfordlhotka** [comment, marked as answer]
Is this Blazor wasm? No, you say server-side Blazor.

Hmm. I did some basic testing and in server-side Blazor it appears that an AuthenticationStateProvider serivce is registered in the environment.

Oh, maybe I'm misreading. You are saying that the data portal endpoint - not the server-side Blazor app - is what's failing. So that's an MVC controller that is trying to use the Blazor context manager and failing I bet.

I _think_ this will be solved by https://github.com/MarimerLLC/csla/issues/2724

**Reply by vijaygill** [reply]
@rockfordlhotka ah yes. My mistake, I was not clear enough. The DataPortal App Server is a WebAPI application hosting the DataPortal. In my test solution, I have this 

1. ASP Core WebAPI application as Data Portal App Server
2. Blazor WASM client (using /api/dataportaltext endpoint of the app-server)
3. Blazor (Server-side) client (/api/dataportal endpoint of the app-server)

I am unable to start the 1st application here (the App Server). 

After having a look at #2724, it seems the solution will cover my scenario.

Thanks!

**Reply by rockfordlhotka** [reply]
Does your app server reference the Csla.Blazor package? That shouldn't be necessary to support a Blazor wasm client, and not referencing that package should resolve your issue (b/c then CSLA will use the aspnetcore context manager).

#2724 will resolve the case where your SSB site _is also hosting the dataportal endpoint_ or other non-SSB web pages or services. It should have no impact on a totally separate webapi website, which wouldn't be referencing the Csla.Blazor package.

Another possible scenario here, is that _maybe_ Blazor wasm doesn't register a default AuthenticationStateProvider like SSB. Are you sure the exception is happening on the app server, or is it maybe happening on the Blazor wasm client?

**Reply by TheCakeMonster** [reply]
Yeah, it does sound like project 1 - your ASP Core WebAPI application acting as the Data Portal App Server - is referencing Csla.Blazor. If that assembly is referenced then the Blazor context manager is used, which is not right for that project/application.

If you drop that reference - which as Rocky says, should not be required for that project - and instead reference Csla.AspNetCore, then it will switch to using the ASP.NET Core implementation of the context manager.

There is an alternative to try for the time being; as part of your config of Csla you could force the context manager being used to the one from Csla.AspNetCore using the new RegisterContextManager<T> config method, something like:

```csharp
services.AddCsla(options => options.RegisterContextManager<Csla.AspNetCore.ApplicationContextManager>())
```
This line of code forces the use of a specific context manager of your choice, rather than letting CSLA choose based upon references.

**Reply by vijaygill** [reply]
> Does your app server reference the Csla.Blazor package? That shouldn't be necessary to support a Blazor wasm client, and not referencing that package should resolve your issue (b/c then CSLA will use the aspnetcore context manager).
> 
> #2724 will resolve the case where your SSB site _is also hosting the dataportal endpoint_ or other non-SSB web pages or services. It should have no impact on a totally separate webapi website, which wouldn't be referencing the Csla.Blazor package.
> 
> Another possible scenario here, is that _maybe_ Blazor wasm doesn't register a default AuthenticationStateProvider like SSB. Are you sure the exception is happening on the app server, or is it maybe happening on the Blazor wasm client?

Hi @rockfordlhotka, yes, you are right! In my test solution I had a project for "Infrastructure" which has some common stuff for both UI and CSLA. By virtue of that project being referenced by AppServer, Csla.Blazor got indirectly referenced in AppServer too. I split that infrastructure project into two - UI specific (i.e. using Csla.Blazor) and CSLA specific (not using Csla.Blazor), Now AppServer references only the CSLA specific infrastructure project and everything is working fine again!

Apologies for all the noise generated due to this mistake on my part.

Thanks again for your help.

**Reply by vijaygill** [reply]
> Yeah, it does sound like project 1 - your ASP Core WebAPI application acting as the Data Portal App Server - is referencing Csla.Blazor. If that assembly is referenced then the Blazor context manager is used, which is not right for that project/application.
> 
> If you drop that reference - which as Rocky says, should not be required for that project - and instead reference Csla.AspNetCore, then it will switch to using the ASP.NET Core implementation of the context manager.
> 
> There is an alternative to try for the time being; as part of your config of Csla you could force the context manager being used to the one from Csla.AspNetCore using the new RegisterContextManager config method, something like:
> 
> ```cs
> services.AddCsla(options => options.RegisterContextManager<Csla.AspNetCore.ApplicationContextManager>())
> ```
> 
> This line of code forces the use of a specific context manager of your choice, rather than letting CSLA choose based upon references.

Hi @TheCakeMonster, it seems you and I were typing our messages at the same time! You and Rocky pointed out the exact error / mistake I had. It's all sorted and working now. Thanks for your input too!

**Reply by TheCakeMonster** [reply]
Yep, so it seems. Glad you got it sorted.
